documentation/field_parameters.sage

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env sage
+
+# This file recomputes the values in src/fp.rs for each FFT-friendly finite
+# field.
+
+import pprint
+
+
+class Field:
+    # The name of the field.
+    name: str
+
+    # The prime modulus that defines the field.
+    modulus: Integer
+
+    # A generator element that generates a large subgroup with an order that's
+    # a power of two. This is _not_ in Montgomery representation.
+    generator_element: Integer
+
+    # The base 2 logarithm of the order of the FFT-friendly multiplicative
+    # subgroup. The generator element will be a 2^num_roots-th root of unity.
+    num_roots: Integer
+
+    def __init__(self, name, modulus, generator_element):
+        assert is_prime(modulus)
+        self.name = name
+        self.modulus = modulus
+        self.generator_element = generator_element
+
+        self.num_roots = None
+        for (prime, power) in factor(modulus - 1):
+            if prime == 2:
+                self.num_roots = power
+                break
+        else:
+            raise Exception(
+                "Multiplicative subgroup order is not a multiple of two"
+            )
+
+    def mu(self):
+        """
+        Computes mu, a constant used during multiplication. It is defined by
+        mu = (-p)^-1 mod r, where r is the modulus implicitly used in wrapping
+        machine word operations.
+        """
+        r = 2 ^ 64
+        return (-self.modulus).inverse_mod(r)
+
+    def r2(self):
+        """
+        Computes R^2 mod p. This constant is used when converting into
+        Montgomery representation. R is the machine word-friendly modulus
+        used in the Montgomery representation.
+        """
+        R = 2 ^ 128
+        return R ^ 2 % self.modulus
+
+    def to_montgomery(self, element):
+        """
+        Transforms an element into its Montgomery representation.
+        """
+        R = 2 ^ 128
+        return element * R % self.modulus
+
+    def bit_mask(self):
+        """
+        An integer with the same bit length as the prime modulus, but with all
+        bits set.
+        """
+        return 2 ^ (self.modulus.nbits()) - 1
+
+    def roots(self):
+        """
+        Returns a list of roots of unity, in Montgomery representation. The
+        value at index i is a 2^i-th root of unity. Note that the first array
+        element will thus be the Montgomery representation of one.
+        """
+        return [
+            self.to_montgomery(
+                pow(
+                    self.generator_element,
+                    2 ^ (self.num_roots - i),
+                    self.modulus,
+                )
+            )
+            for i in range(min(self.num_roots, 20) + 1)
+        ]
+
+
+FIELDS = [
+    Field(
+        "FieldPrio2",
+        2 ^ 20 * 4095 + 1,
+        3925978153,
+    ),
+    Field(
+        "Field64",
+        2 ^ 32 * 4294967295 + 1,
+        pow(7, 4294967295, 2 ^ 32 * 4294967295 + 1),
```

Comment:
How were the generators originally chosen (I get that they're defined in the spec)? Just some arbitrary value [that fit this criteria](https://www.ietf.org/archive/id/draft-irtf-cfrg-vdaf-07.html#name-fft-friendly-fields)?
> To be FFT-friendly, the order of this subgroup MUST be a power of 2. In addition, the size of the subgroup dictates how large interpolated polynomials can be. It is RECOMMENDED that a generator is chosen with order at least 2^20.

Reply:
I wasn't able to figure out how the generator used with `FieldPrio2` was selected, but it has the right order, so it works. For the other two, the definitions use a neat trick wherein they start with an arbitrary group member with an order that's too high, and then knock off the unneeded cofactors through exponentiation until they get an element with the desired order. (If $g$ has order 6, then by the definition of the order of an element, $g^6 = e$, where $e$ is the identity, and $g^2 \neq e$ and $g^3 \neq e$. Then we can set $h = g^3$, and by substituting we know $h^2 = (g^3)^2 = g^6 = e$. By assumption, $h = g^3 \neq e$, thus $h$ has order two.) Choosing an arbitrary group element in a group this big has a large chance of selecting one with an order equal to the group's order, and 7 worked out as a starting place for both cases. If it hadn't, we could try again with something else.

```
sage: Field64 = GF(2^32 * 4294967295 + 1)
sage: factor(Field64.unit_group().order())
2^32 * 3 * 5 * 17 * 257 * 65537
sage: factor(Field64(7).multiplicative_order())
2^32 * 3 * 5 * 17 * 257 * 65537
sage: factor((Field64(7) ^ 4294967295).multiplicative_order())
2^32
sage: Field128 = GF(2^66 * 4611686018427387897 + 1)
sage: factor(Field128.unit_group().order())
2^66 * 3 * 3491 * 440340496364689
sage: factor(Field128(7).multiplicative_order())
2^66 * 3 * 3491 * 440340496364689
sage: factor((Field128(7) ^ 4611686018427387897).multiplicative_order())
2^66
```